prototype/storage_api.py

```diff
@@ -0,0 +1,175 @@
+
+
+@app.route("/")
+def root():
+    return {"message": "Storage api for pipelined mmif files"}
+
+
+@app.route("/upload_mmif/", methods=["POST"])
+def upload_mmif():
+    body = request.get_data(as_text=True)
+    # read local storage directory from config.yml
+    with open('config.yml', 'r') as file:
+        config = yaml.safe_load(file)
+    directory = config['storage_dir']
+    mmif = Mmif(body)
+    # get guid from location
+    # document = body.[0]['properties']['location'].split('/')[2].split('.')[0]
+    document = mmif.documents['d1']['properties'].location.split('/')[2].split('.')[0]
+    # append '.mmif' to guid
+    document = document + '.mmif'
+    # IMPORTANT: In order to enable directory creation after this loop and also store each parameter
+    # dictionary in its proper directory, I create a dictionary to associate the current path level with
+    # its param dict. After this loop, I create the dirs and then iterate through this dictionary to
+    # place the param dicts in their proper spots.
+    param_path_dict = {}
+    for view in mmif.views:
+        # this should return the back half of the app url, so just app name and version number
+        subdir_list = view.metadata.app.split('/')[3:]
+        # create path string for this view
+        view_path = os.path.join('', *subdir_list)
+        # now we want to convert the parameter dictionary to a string and then hash it.
+        # this hash will be the name of another subdirectory.
+        try:
+            param_dict = view.metadata["parameters"]
+            param_list = ['='.join(pair) for pair in param_dict.items()]
+            param_list.sort()
+            param_string = ','.join(param_list)
+        except KeyError:
+            param_dict = ""
+            param_string = ""
+        # hash the (sorted and concatenated list of params) string and join with path
+        # NOTE: this is *not* for security purposes, so the usage of md5 is not an issue.
+        param_hash = hashlib.md5(param_string.encode('utf-8')).hexdigest()
+        view_path = os.path.join(view_path, param_hash)
+        # check if this is a duplicate view. if it is, skip the current view.
+        # NOTE: duplicate views are those with the same app, version number, AND parameter dict.
+        if view_path in directory:
+            continue
+        # create path by joining directory with the current view path
+        directory = os.path.join(directory, view_path)
+        # now that we know it's not a duplicate view and we have the proper path location, we
+        # store it and the associated param dict inside param_path_dict.
+        param_path_dict[directory] = param_dict
+    # we have finished looping through the views. now time to create the directories
+    # and dump the param dicts
+    os.makedirs(directory, exist_ok=True)
+    for path in param_path_dict:
+        file_path = os.path.join(path, 'parameters.json')
```

Comment:
I tried this out, and figured that it would be more convenient for clients if we just store `.mmif` files only in this directory, and have an additional api to return the entire absolute path of the directory (just like how `baapb://` is resolved), so that when we run some code that uses pre-stored mmif files on the same machine where those files are, the client doesn't have to "download" the mmif files, but directly can access the directory (if exists) and use the files under.

Reply:
And if we store the parameter files in the "appversion" directory, and keeping only mmif files in the "hash" directory, clients can just take `*` glob to read all mmif files, without worrying about any additional metadata-like files. 
So instead of
```
storage-root/
├── app1
│   ├── v1
│   │   ├── hash1
│   │   │   ├── guid1.mmif
│   │   │   └── params.json
│   │   ├── hash2
│   │   │   ├── guid1.mmif
│   │   │   └── params.json
│   │   └── hash3
│   │       ├── guid1.mmif
│   │       └── params.json
│   └── v2
│       └── hash1
│           ├── guid1.mmif
│           └── params.json
└── app2
    ├── v1
    │   └── hash1
    │       ├── guid1.mmif
    │       └── params.json
    └── v2
        └── hash1
            ├── guid1.mmif
            └── params.json
```
, 
```
storage-root/
├── app1
│   ├── v1
│   │   ├── hash1
│   │   │   └── guid1.mmif
│   │   ├── hash1.json
│   │   ├── hash2
│   │   │   └── guid1.mmif
│   │   ├── hash2.json
│   │   ├── hash3
│   │   │   └── guid1.mmif
│   │   └── hash3.json
│   └── v2
│       ├── hash1
│       │   └── guid1.mmif
│       └── hash1.json
└── app2
    ├── v1
    │   ├── hash1
    │   │   └── guid1.mmif
    │   └── hash1.json
    └── v2
        ├── hash1
        │   └── guid1.mmif
        └── hash1.json
```